designs/replicated-pv/mayastor/qos.md

```diff
@@ -0,0 +1,202 @@
+
+#### Story 3: Adding QoS to Non-QoS StorageClass
+
+As a user, I want to add QoS annotations to a PVC that uses a StorageClass without QoS limits so that I can apply performance controls to volumes that didn't originally have them.
+
+**StorageClass without QoS:**
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: mayastor
+provisioner: io.openebs.csi-mayastor
+parameters:
+  protocol: "nvmf"
+  repl: "2"
+```
+
+**PVC adding QoS via annotations:**
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: analytics-storage
+  annotations:
+    openebs.io/qos-iops-limit: "3000"
+    openebs.io/qos-bandwidth-limit-mb: "150"
+spec:
+  storageClassName: mayastor-basic
+  ...
+```
+
+### Implementation Details
+
+**CSI Integration:**
+- CSI driver processes StorageClass parameters for initial QoS configuration
+- SPDK QoS integration via existing JsonGrpc proxy
+- VolumeSpec integration for persistence
+
+**QoS Parameters:**
+- SPDK rate limit types (`rw_ios_per_sec`, `rw_mbytes_per_sec`, `r_mbytes_per_sec`, `w_mbytes_per_sec`) mapped to new StorageClass parameters and PVC annotations
+- Annotation names match StorageClass parameter names using `openebs.io/` namespace prefix
+- `qos-iops-limit` / `openebs.io/qos-iops-limit` - IOPS rate limit, becomes `rw_ios_per_sec`
+- `qos-bandwidth-limit-mb` / `openebs.io/qos-bandwidth-limit-mb` - Combined read/write bandwidth limit, becomes `rw_mbytes_per_sec`
+- `qos-read-bandwidth-limit-mb` / `openebs.io/qos-read-bandwidth-limit-mb` - Read-only bandwidth limit, becomes `r_mbytes_per_sec`
+- `qos-write-bandwidth-limit-mb` / `openebs.io/qos-write-bandwidth-limit-mb` - Write-only bandwidth limit, becomes `w_mbytes_per_sec`
+
+**PVC Annotation Monitoring:**
+- Dedicated PVC watcher added to CSI driver monitors annotation changes on existing PVCs
+- Detects QoS annotation additions, modifications, and removals
+- Validates parameters before applying changes to live volumes
+- Applies changes to live volumes without restart via JsonGrpc → SPDK RPC calls
+- Updates VolumeSpec persistence layer to maintain consistency
+- Reports validation errors and operation status via PVC status field
+
+**Status Reporting:**
+- Applied QoS limits added to volume status when configured for operational visibility
+- Validation errors and conflicts reported in PVC status field
+- Status reflects actual SPDK-applied limits regardless of source (StorageClass or PVC annotations)
+
+**Validation and Constraints:**
+- Parameter validation ensures non-negative values and meets SPDK minimum limits
+- QoS applied at nexus level (published volumes only)
+- SPDK minimum limits: 1000 IOPS, 10 MB/s bandwidth
+- Units: IOPS (operations/second), bandwidth in megabytes/second
+
+### Risks and Mitigations
+
+**Resource Contention:**
+- Risk: Multiple volumes with high QoS limits could exceed underlying storage capabilities, causing system-wide performance degradation despite individual volume limits.
```

Comment:
For this we should expose accrued QoS on a given node or pool, allowing users to check if any resource is potentially oversubscribed?

Reply:
Hm, I don't think we need to expose anything else. I would hope it is a matter monitoring the node itself. But it wouldn't hurt to make users aware of that in the Documentation or FAQ.

Reply:
Yeah that could be a future enhancement.
Would we then take into account as a preference or enforcement? If enforcement then what do do if there's no available bandwidth, treat as a failure?

Reply:
> If enforcement then what do do if there's no available bandwidth, treat as a failure?

We don't really know the proper bandwidth in advance and it might even depend on access patterns. In my mind this is mostly best-effort. You still need monitoring to verify the system is performing as expected.